### PR TITLE
[ChessGame] New features for the bot owner.

### PR DESCRIPTION
**Pull request by AAA3A-AAA3A**
Hello,
I tried to modify the code of your cog to bring new features.
The first one is the fact that the bot owner can close a chess game by doing the command "[p]chess close game channel". The "channel" argument is optional to be able to close a game that is not in the command's channel.
The second is the fact that the owner can start a game between the two specified players with the agreement of course of each player with two reaction messages.
Thanks in advance and have a nice day!

### CODE REVIEW

**Comment by WildStriker** [comment]
Hey @AAA3A-AAA3A,

thanks for your contributions!

I'll review it later today, and when I have a chance to load it up and test it out.

some quick things I notice:

- there is some refactoring needed here
- I think we can reduce a bit of repeated code
    - start game and launch can probably share a "utility" function for the last few steps that are exactly the same
    - rename your launch command (currently `async def start_game`, maybe just `async def launch`?)
    - I also noticed _display_board is defined twice so we will want to remove that
 

close feature looks good for my initial scan :)

File: chessgame/chessgame.py [hunk 1]
```diff
@@ -140,6 +140,138 @@ async def start_game(self, ctx: commands.Context,
 
         await self._display_board(ctx, embed, game)
 
+    async def _display_board(self, ctx: commands.Context, embed: discord.Embed, game: Game):
```
Comment (WildStriker) [inline chessgame/chessgame.py]: this is duplicated, either remove this, or the one below?

File: chessgame/chessgame.py [hunk 2]
```diff
@@ -140,6 +140,138 @@ async def start_game(self, ctx: commands.Context,
...
+        """displays the game board"""
+        board_image = io.BytesIO(game.get_board_image())
+        embed.set_image(url="attachment://board.png")
+        await ctx.send(embed=embed, file=discord.File(board_image, 'board.png'))
+        
+    @commands.is_owner()
+    @chess.command(name='launch', autohelp=False, help=start_help_text())
+    async def start_game(self, ctx: commands.Context,
```
Comment (WildStriker) [inline chessgame/chessgame.py]: rename this function, not to conflict with the existing one

File: chessgame/chessgame.py [hunk 3]
```diff
@@ -140,6 +140,138 @@ async def start_game(self, ctx: commands.Context,
...
+        embed: discord.Embed = discord.Embed()
+        embed.title = "Chess"
+        embed.description = f"Game: {game_name}"
+        embed.add_field(name="Type:", value=game.type, inline=False)
+
+        embed.add_field(name="New Game",
+                        value=f"<@{player_white.id}>'s (White's) turn is first",
+                        inline=False)
+
+        await self._display_board(ctx, embed, game)
```
Comment (WildStriker) [inline chessgame/chessgame.py]: all this is should be exactly the same as start_game

maybe refactor so we have a helper _start_game(ctx, player_black, player_white) function that both these commands can utilize to reduce repeated code

File: chessgame/chessgame.py [hunk 4]
```diff
@@ -140,6 +140,138 @@ async def start_game(self, ctx: commands.Context,
...
+                if not pred.result:
+                    create_game = False
+                    embed.add_field(
+                        name="Response:",
+                        value="One of the players refused the game!")
+            except asyncio.TimeoutError:
+                create_game = False
+                embed.add_field(
+                    name="Timed out:",
+                    value=f"<@{other_player.id}> did not respond in time.")
```
Comment (WildStriker) [inline chessgame/chessgame.py]: I am wondering if we can combine this into one prompt

- that would give both users a chance to respond right away (user 2 might have to wait on user 1 for 9 plus minutes :smile: )
- Could be a bit more user friendly?
    - bonus, I think it would be less code to implement too! (hopefully, if not the same) 




i.e. the prompt could say something like:
>  `bot owner` wishes to start a game between `@user_one` and `@user_two` please respond if you wish to play below

or something to that effect..

then you could can wait for a reaction from any user and validate it is one of the users we care about:
```py
    pred = ReactionPredicate.yes_or_no(
        message)

    users = {player_1.id, player_2.id}
    try:
        while users:
            # wait for anyone to respond
            _reaction, user = await ctx.bot.wait_for("reaction_add", check=pred,  timeout=600)

            # TODO maybe also give the bot owner a chance to cancel out here?
            if user.id in users:
                if pred.result:
                    users.remove(user.id)
                    # TODO update message with first users response
                else:
                    # TODO a user has declined, update message
                    return
    except asyncio.TimeoutError:
        # TODO one of if not both of the players did not respond

    # TODO start the game!
```

**Review by WildStriker — COMMENTED** [review]
It looks like you got some of these already, forgot to submit the review

File: chessgame/chessgame.py (hunk 4 above)
Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: I put a personalized message to explain to the user that he has the possibility to play, without mentioning the owner of the bot.
Maybe remove the bot owner's message from the command?
Maybe allow to specify the channel where the game will start?

File: chessgame/chessgame.py (hunk 3 above)
Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: I didn't do this because I didn't know what the arguments were for the command. Ctx and both players in the end.

Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: Following your post the other time, I had made two of the three changes. As you can see just below your first post, the changes took effect on this PR.

Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: I tried to use the same code for both commands with "await self._start_game(ctx, player, other_player, game_name)", but Python says "local variable 'game_name' referenced before assignment".

Comment (WildStriker) [reply chessgame/chessgame.py]: Sorry, I obviously missed some things to include in my example

You will need to pass more variables over.  The complete signature should look something like this:

```py

    async def start_game(...):
        # ... prompts / initial set up
        await self._start_game(ctx, ctx.author, other_player, game_name, game_type)

    async def launch_game(...):
        # ... prompts / initial set up
        await self._start_game(ctx, player, other_player, game_name, game_type)

    # ... then definition of the helper function:
    async def _start_game(self, ctx: commands.Context,
                 player_black: discord.Member, player_white: discord.Member,
                 game_name: str = None, game_type: str = None):

        # get games config
        games = await self._get_games(ctx.channel)
        if not games:
            games = {}
        
        # remove these, no longer needed
        # player_black = ctx.author
        # player_white = other_player

        # ... the rest is exactly the same
        # ...
        # await self._display_board(ctx, embed, game)

```

**Note: There is no need for setting up the player_black and player_white variables in the helper function so remove them (line 233 / 234) these will should be part of the new functions parameters


Hope this is clear, sorry if it is confusing.  The comments are meant to keep this example short all code that is not duplicated (or not the above note) should still be intact :smile: 

So Basically:
- create a new function and copy all code from the `# get games config` comment to `await self._display_board(ctx, embed, game)`
- then remove `player_black` and `player_white` initialization in the new function
- remove the duplicated code from `start_game` and `launch_game` and call the new function

File: chessgame/chessgame.py (hunk 4 above)
Comment (WildStriker) [reply chessgame/chessgame.py]: Not sure I follow.  The bot owner does not have to be in the prompt, at the very least it does not have to be an @mention

- Technically we can see who issued the command so it doesn't really add any useful information.

I was just thinking we can use just one prompt so that the users may respond in any order

- It would probably be nice add an additional check if the bot owner wants to cancels the game before it starts
    -  Just in case they selected the wrong people

The prompt could simply be two fields with a description
- Description: Would you like to start a game of chess?
- Additional fields can be listed below to hold the users responses (should be 4 possible states):
    - User Did not Respond Yet
    - User Accepts
    - User Declines
    - Timed Out (no response at all)

e.g. for the additional fields it could be displayed like so:
| @user_1  | @user_2  |         
| ------------- |:-------------:|
| Please Respond Below | Has declined! |

| @user_1  | @user_2  |         
| ------------- |:-------------:|
| Has accepted! | Did Not Respond! (Timed out) |

This is just an example, doesn't have to look like this.

And possibly if there is a way for owner to cancel, we can clear out the existing fields and put cancelled by owner, or just simply remove the message

Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: In your previous post, you offered a sample post that said the bot owner had submitted this part. I preferred that the message not even say who started a game.

File: chessgame/chessgame.py (hunk 3 above)
Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: I'm not at all good at Python. I will try again, even if I do not understand some things.

File: chessgame/chessgame.py (hunk 4 above)
Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: A confirmation message for the bot owner. Maybe also a forbidden panel reaction to allow the owner to cancel a game request or even directly on the game?

Comment (WildStriker) [reply chessgame/chessgame.py]: Yeah, to be clear I am open to suggestions on how you would like to word the message.

I did add "bot owner" in my previous example, but it is not necessary to do so and you have some creative freedom on how to word it is all I meant.

My main point is I wanted to see just one prompt but the way it is currently implemented there is two messages.

I'll play around with the idea if you need a full example but the example code I mentioned previously has most of it implemented (besides the actual message and editing of the user response fields)

File: chessgame/chessgame.py (hunk 3 above)
Comment (WildStriker) [reply chessgame/chessgame.py]: If you need a hand let me know :)

maybe we can discuss this easier in discord?

Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: Yes.

File: chessgame/chessgame.py (hunk 4 above)
Comment (AAA3A-AAA3A) [reply chessgame/chessgame.py]: I agree with you for a single confirmation message.

**Comment by AAA3A-AAA3A** [comment]
I'm coming back to this PR because I haven't had any returns since 28 October (1 month and 20 days). Do you have any new proposals for changes?  How to do the 2 things that were left pending and that I could not implement? Thank you in advance for your reply.
PS: In the meantime, I have started my own cogs and I am working on my 5th (SimpleSanction). So I won't be available much, but I know more about python and cogs for the redbot.

**Comment by WildStriker** [comment]
Hi @AAA3A-AAA3A,

I made some changes on top of this pull request.

If you can allow me to make changes I can probably push it here.. (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

Alternatively I might need to do a separate PR on your fork so you can merge it in

I removed the repeated code so both start_game and launch_game uses mostly the same logic.

I added a prompt that allows both users respond in any order for launch_game

only pending item I think will be to do a rebase / deal with conflict to the changes I made a week ago to the master branch.

**Comment by AAA3A-AAA3A** [comment]
Hi @WildStriker,

I usually allow fork changes like in your link. The "Allons édits by maintainers" box is already checked for this PR. If you can't do it, do a PR and I would accept it in a heartbeat.

I saw below this page that there was a conflict, but I was not sure that it did not already exist before. What are your modifications? I haven't updated my cogs for a long time so as not to lose my changes intended for PRs.

When you have modified the code of this PR, I will add myself (it has already been done) the code for a DISABLED check of the bots before each game.

If you want to discuss more easily, you can find me on both Red servers: the main one and the cogs server.

Thanks you in advance.

**Comment by WildStriker** [comment]
Hi @AAA3A-AAA3A,

Sorry my mistake, it was a token issue.  I pushed my changes I previously made.

**Comment by WildStriker** [comment]
Changes have been merged. Fixed a few minor issues as well.

Thanks!

**Comment by AAA3A-AAA3A** [comment]
Thanks to you for the merge and for the help in the code, instead!
